GPU/Common/SoftwareTransformCommon.cpp

```diff
@@ -674,15 +686,21 @@ void SoftwareTransform::CalcCullParams(float &minZValue, float &maxZValue) {
 		std::swap(minZValue, maxZValue);
 }
 
-void SoftwareTransform::ExpandRectangles(int vertexCount, int &maxIndex, u16 *&inds, const TransformedVertex *transformed, TransformedVertex *transformedExpanded, int &numTrans, bool throughmode) {
+bool SoftwareTransform::ExpandRectangles(int vertexCount, int &maxIndex, u16 *inds, int &indsOffset, int indexBufferSize, const TransformedVertex *transformed, TransformedVertex *transformedExpanded, int &numTrans, bool throughmode) {
+	// Before we start, do a sanity check - does the output fit?
+	if ((vertexCount / 2) * 6 > indexBufferSize - indsOffset) {
+		// Won't fit, kill the draw.
+		return false;
+	}
+
 	// Rectangles always need 2 vertices, disregard the last one if there's an odd number.
 	vertexCount = vertexCount & ~1;
 	numTrans = 0;
 	TransformedVertex *trans = &transformedExpanded[0];
 
-	const u16 *indsIn = (const u16 *)inds;
-	u16 *newInds = inds + vertexCount;
-	u16 *indsOut = newInds;
+	const u16 *indsIn = (const u16 *)(inds + indsOffset);
+	int newIndsOffset = indsOffset + vertexCount;
```

Comment:
Won't this not add enough space to not overlap?  Not sure if maybe it's still safe.  Before it was `(u16 *)x + y` aka `x + y * 2` and now it's `x + y`.  Also in other places here.

-[Unknown]

Reply:
Oh duh, sloppy of me to miss. My test cases still worked somehow though..

Reply:
Actually hm, indsOffset is in 16-bit units here, since we just after add it to a u16*.  But I think there might be some unclarity here, will check it all.

Reply:
Oh, I thought one of these had been changed to a u8 because of the cast, I guess.  My mistake.

-[Unknown]

Reply:
Well, something was indeed wrong and I had to revert, but maybe not here. Will redo this later.